frontend/i_types.py

```diff
@@ -0,0 +1,274 @@
+from abc import ABCMeta, abstractmethod
+
+class Type(metaclass=ABCMeta):
+    """This class is the top-parent of all possible inferred types of a Python program.
+    Every type has its own class that inherits this class.
+    """
+
+    @abstractmethod
+    def is_subtype(self, t):
+        """Check if this type is a subtype of the parameter type.
+
+        Arguments:
+            type (Type): The type to check the subtype relationship against.
+        """
+        pass
+
+
+    @abstractmethod
+    def get_name(self):
+        """Get the name of the type"""
+        pass
+
+    # TODO: Implement __eq__ and __hash__ for all types, for correct set behaviour
+
+
+class TNone(Type):
+
+    def is_subtype(self, t):
+        return True
+
+    def get_name(self):
+        return "None"
+
+
+class TBool(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, (TBool, TInt, TFloat))
+
+    def get_name(self):
+        return "Bool"
+
+
+class TInt(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, (TInt, TFloat))
+	
+    def get_name(self):
+        return "Int"
+
+
+class TFloat(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, TFloat)
+
+    def get_name(self):
+        return "Float"
+
+
+class TString(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, TString)
+
+    def get_name(self):
+        return "String"
+
+
+class TBytesString(Type):
+
+    def is_subtype(self, t):
+        return isinstance(t, TBytesString)
+
+    def get_name(self):
+        return "ByteString"
+
+
+class TList(Type):
+    """Type given to homogeneous lists.
+
+    Attributes:
+        type (Type): Type of the list elements
+    """
+
+    def __init__(self, t):
+        self.type = t
+
+    def is_subtype(self, t):
+        return isinstance(t, TList) and self.type.is_subtype(t.type)
```

Comment:
Normally, you would want the type argument in both types to be identical. List[str] should not be a subtype of List[object]. 
Also: Note that you essentially have the same code here as in is_subtype of TSet. You could avoid that if you had something like a GenericType class.
Also: The way it is right now, it's impossible to have a user-defined class that extends list. Or rather, if there was a class like this, then this is_subtype implementation is incorrect, because the isinstance(t, TList) check would fail, even though it shouldn't.